aws/rust-runtime/aws-auth/src/refresh.rs

```diff
@@ -0,0 +1,294 @@
+                .name("RefreshingCredentialProvider-executor".into())
+                .spawn(move || {
+                    RefreshingCredentialProvider::executor_thread(executor_signal, executor_inner);
+                })
+                .map_err(|err| Error::ThreadSpawnFailed(Box::new(err)))
+                .and_then(|joiner| joiner.join().map_err(|_| Error::ExecutorThreadFailed));
+            if let Err(err) = result {
+                error!("RefreshingCredentialProvider executor failed: {}", err);
+                thread::sleep(EXECUTOR_PANIC_BACKOFF);
+            }
+        }
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    fn executor_thread(signal: RwCell<RefreshLoopState>, inner: Inner) {
+        let runtime = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .expect("failed to start tokio runtime");
+        runtime.block_on(async move {
+            // TODO: Take a name argument from users to indicate which credentials are being refreshed?
+            let span = trace_span!("credential_refresh_loop");
+            let _enter = span.enter();
+
+            info!("Starting credential refresh loop");
+            Self::refresh_loop(signal, inner)
+                .await
+                .expect("refresh loop failed");
+        });
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    async fn refresh_loop(signal: RwCell<RefreshLoopState>, inner: Inner) -> Result<(), Error> {
+        while let RefreshLoopState::Running = signal.get() {
+            while let Some(sleep_duration) = inner.next_refresh_sleep_time(SystemTime::now()) {
+                tokio::time::sleep(sleep_duration).await;
+            }
+
+            info!("Refreshing credentials");
+            let timeout_future = tokio::time::sleep(REFRESH_TIMEOUT);
+            let refresh_future = inner.refresh();
+            tokio::pin!(timeout_future, refresh_future);
+            loop {
+                // In the event of timeout or failure, the sleep in the next iteration will be
+                // skipped because the credential expiration time won't have changed.
+                // TODO: Should we have backoff for both timeout and refresh failures, or just refresh failures?
+                tokio::select! {
+                    _ = timeout_future => {
+                        error!("Credential refresh timed out after {} seconds", REFRESH_TIMEOUT.as_secs());
+                        break;
+                    }
+                    result = refresh_future => {
+                        if let Err(err) = result {
+                            error!("Failed to refresh credentials: {}", err);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl ProvideCredentials for RefreshingCredentialProvider {
+    fn provide_credentials(&self) -> Result<Credentials, CredentialsError> {
```

Comment:
I wonder if we actually want some sort of tri-state here where if there is an error loading credentials, we actually set the error so we can return it